### ISSUE TITLE
Inconsistent (de)serialization state with use `calcTransformMatrix` and `qrDecompose`

### ISSUE THREAD

**Issue by vitonsky** [fabricjs/fabric.js#7620]
Hi. I need to clone objects and i should serialize the objects for my purposes.
When i try to copy the object which is flipped, i got unexpected transformation (see below).
How we can fix it?

## Version
4.6.0

## Test Case
https://jsfiddle.net/k3wof27e/

## Information about environment
Nodejs or browser? Browser
Which browsers? All

## Steps to reproduce
- Open https://jsfiddle.net/k3wof27e/
- Flip rect manually
- Press button to clone object

## Expected Behavior
Cloned object is have same rotation state and position as original object

## Actual Behavior
Cloned object is not flipped. It's looks like it's flipped, because transform controls are flipped, but also object rotated to 180deg.

**Comment by asturur** [comment]
this is a known issue for which we don't have a solution.
Don't use qrDecompose before cloning.
Is unclear why you have to do that

**Comment by ShaMan123** [comment]
#6941

**Comment by asturur** [comment]
i think his issue isn't the wobbling but more the swap between flipX + flipY with a rotarion.

**Comment by vitonsky** [comment]
@asturur 
> Don't use qrDecompose before cloning.

I use it after. It did required to get absolute object coordinates instead coordinates relative group.
At this time i can do without it, but why this pair of utils work incorrectly together for this case? Is it planning to fix in future?

**Comment by asturur** [comment]
is not that they do not work correctly together.
extracting angle/scale from a generic matrix is absolutely a lossy process.
6 number = 8 values ( left, top, scaleX, scaleY, angle, skewX, flipX, flipY ), and a bunch of trigronometry in the middle.
Something get lost and i think there is no way to fix it.
If someone knows how to fix it, great.

**Comment by ShaMan123** [comment]
I terms of linear algebra this should be solvable.
I don't understand why it's inaccurate (read about the floating point issue), nor do I understand the calculation itself...

**Comment by asturur** [comment]
when i say lossy i mean that properties are mixed up and you can extract one of the many solutions that bring to that mixup, but not the exact one.

If you take a paper with some drawing, and you flip it vertically and then horizontally, you will see that is rotated by 180 degree. once i give you a rotated picture, you don't know if i got there by flipping or by rotating.
Same for this process here:

![image](https://user-images.githubusercontent.com/1194048/150655556-29df6eea-e63a-4d37-b152-0373f3ed2438.png)
this image has skewX + skewY

![image](https://user-images.githubusercontent.com/1194048/150655562-5ec64474-c588-490e-9f0b-72651a600f2f.png)
this image has skewX + angle

but the end result is the same. qrDecompose start from the end result.

For the numbers accuracy, javascript has holes, as hany 64bit floating point number representation
the easiest example is:
![image](https://user-images.githubusercontent.com/1194048/150655752-7149670d-24b9-4c6a-a2ba-98136a529084.png)

64bits give you 2^64 combinations of values
Considering that you can represent integer number with precision ( all of them ) up to 999999999999999 ( 15 digits ) if you try to type something bigger in the console you get other numbers.
now that number is a bit bigger than 2^47 but just to give an idea around 2^47 combiation of 2^64 are used for this.
you have left 64-47 bits so 17 bits. This means that each of those number has around 17 bits extra to represent the decimals. So for each of those numbers you have around 128k decimal values. All the other are missing.
Now this is explained really in a simplistic way and the complexity is reduced to the bone, how the floating point works is a bit complex to remember and explain, but as a matter of fact for a lot of fractions, the calculation is approximated.

**Comment by ShaMan123** [comment]
So in theory if we split up each number into a whole number and the remaining fraction the precision will be better?

**Comment by ShaMan123** [comment]
What I saw is when I have both skewX and skewY this happens. If only one of them accuracy remains.

**Comment by asturur** [comment]
> So in theory if we split up each number into a whole number and the remaining fraction the precision will be better?

No. Floating point has internal things, is not that simple. Is the best precision you can get with those 64bits, we can't improve it.

@ShaMan123 your math skills are probably better than mine.

To write qrDecompose i learned from this video here:
https://www.youtube.com/watch?v=51MRHjKSbtk

If you find something better great!

**Comment by asturur** [comment]
![image](https://user-images.githubusercontent.com/1194048/150656615-9ab7f4d5-5ab5-42e7-8f9b-0b556d6b471b.png)

39,383 views of which probably 383 are mine.

**Comment by ShaMan123** [comment]
Haha you made me laugh

**Comment by ShaMan123** [comment]
heavy stuff
https://en.wikipedia.org/wiki/QR_decomposition

**Comment by ShaMan123** [comment]
I wouldn't be sure that your math skills are inferior to mine after trying to understand what goes where and for what reason...

**Comment by ShaMan123** [comment]
Could you explain why QR decomposition is needed?
I need more context.
In linear algebra context I do understand that there's a need to collapse 8-9 dimensions into 6.
But that's all I've managed to understand. And maybe we can calculate the kernel instead. But I'm just shooting things out my waist.

**Comment by asturur** [comment]
qrDecomposition was made to solve the issues of rotated objects in groups that got scaled and then un-grouped. is the only way i found to split a generic transformation into a set of parameters a human developer can understand. scale, rotation and skewing.

**Comment by vitonsky** [comment]
@asturur i think we should consider the `flipX`/`flipY` properties while creation matrix.

For example, we can use negative values to describe flipped objects and positive for not flipped.

What you think about?

**Comment by asturur** [comment]
is not that simple, rotation can swap the scale signs anyway.

**Comment by vitonsky** [comment]
> is not that simple, rotation can swap the scale signs anyway.

What you about? I not so understand.

Main idea - do it on matrix calculation stage, when we have full info about object. If current encoding is already use negative values for something and we can't use it that, so let's introduce parameter `encodingMethod` for `calcTransformMatrix`.

For example, we can limit rotation only 0-360deg and instead 361 use 1 if necessary

**Comment by ShaMan123** [comment]
I am trying to understand the reason why QR decomposition is needed.
If I am right it is intended to represent the transform matrix by eigenvectors/eigenvalues. Is this correct?
I am guessing it is needed so that the translate values remain the same.
Is this correct?
For any other reason I don't understand why an eigen representation will be more helpful than another (meaning I don't see a need for this calculation)

**Comment by ShaMan123** [comment]
Good links:
https://farazzshaikh.medium.com/affine-transformations-pan-zoom-skew-96a3adf38eb2
https://medium.com/hipster-color-science/computing-2d-affine-transformations-using-only-matrix-multiplication-2ccb31b52181
https://www.w3resource.com/html5-canvas/html5-canvas-matrix-transforms.php

**Comment by ShaMan123** [comment]
https://github.com/ShaMan123/fabric.js/blob/5a3164c0cf16dd9e616077d6849a0b167c85502f/src/util/misc.js#L769

**Comment by ShaMan123** [comment]
I have found an implementation of QRDecomposition [here](https://github.com/chrvadala/transformation-matrix/blob/dfcda883aecbaeb10c8bde42d375a97878b5276c/src/decompose.js#L16) we can test.

**Comment by ShaMan123** [comment]
I believe #8767 is a way out of most of this trouble

**Comment by zhe-he** [comment]
> #8767
I encountered a similar issue, and I'm curious why we can't directly use a negative value of scaleX instead of flipX. This way, there would be one less dimension in the matrix. When I want to check whether an object has flipX and flipY, I can check if scaleX and scaleY are negative. I know there's a built-in method applyTransformToObject, which after execution, forces flipX and flipY to be false. I'm puzzled why it can't be split into two methods. The first one is to read the values of obj's flipX and flipY when executing applyTransformToObject and not change them. The second method is to modify obj's flipX and flipY. Also, when fabric internally draws, it rotates 180 degrees based on flipX. I'm particularly puzzled, if scaleX is negative, can we avoid the judgment and directly draw? In the end, each shape has the following properties: width, height, flipX, flipY, and matrix. The matrix is parsed into left, top, scale X, scaleY, skewX, skewY.
